### PR TITLE
fix: only show severity total from filter flags 

### LINKED ISSUE #533: Omit counts for severity levels that are excluded from output
**Description**

If I use the `--severity` option, Trivy correctly only shows vulnerabilities for matching severities. It still shows a total, which is always 0, for other severities, which is misleading. 

**What did you expect to happen?**

Only include the totals for severities that are being counted. For example, if `--severities=HIGH,CRITICAL` then I think the totals output should be like this: 
```
Total: 1 (HIGH: 1, CRITICAL: 0)
``` 
omitting the (misleading) 0 counts for unknown, low and medium. 

**What happened instead?**

```
trivy repo --severity=HIGH,CRITICAL https://github.com/tootsuite/mastodon                                                                 
Enumerating objects: 38265, done.
Counting objects: 100% (38265/38265), done.
Compressing objects: 100% (20176/20176), done.
Total 38265 (delta 22474), reused 31711 (delta 17086), pack-reused 0
2020-06-18T14:21:41.418+0100	WARN	OS is not detected and vulnerabilities in OS packages are not detected.
2020-06-18T14:21:41.418+0100	INFO	Detecting yarn vulnerabilities...
2020-06-18T14:21:41.430+0100	INFO	Detecting bundler vulnerabilities...

Gemfile.lock
============
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)


yarn.lock
=========
Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)

+---------+------------------+----------+-------------------+---------------+--------------------------------+
| LIBRARY | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |             TITLE              |
+---------+------------------+----------+-------------------+---------------+--------------------------------+
| lodash  | NSWG-ECO-516     | HIGH     | 4.17.15           |               | Allocation of Resources        |
|         |                  |          |                   |               | Without Limits or Throttling   |
+---------+------------------+----------+-------------------+---------------+--------------------------------+
```

**Output of `trivy -v`:**

```
Version: 0.9.1
Vulnerability DB:
  Type: Light
  Version: 1
  UpdatedAt: 2020-06-18 12:09:35.068151619 +0000 UTC
  NextUpdate: 2020-06-19 00:09:35.068151319 +0000 UTC
```

**Additional details (base image name, container registry info...):**

### PR DESCRIPTION
**Pull request by h4ckm03d**
Fix #533 

Test running:
1. With `HIGH,CRITICAL` severity
```
go run cmd/trivy/main.go repo --severity=HIGH,CRITICAL https://github.com/tootsuite/mastodon        
Enumerating objects: 38799, done.
Counting objects: 100% (38799/38799), done.
Compressing objects: 100% (20524/20524), done.
Total 38799 (delta 22702), reused 32154 (delta 17257), pack-reused 0
2020-07-14T16:23:17.135+0700    WARN    OS is not detected and vulnerabilities in OS packages are not detected.
2020-07-14T16:23:17.135+0700    INFO    Detecting yarn vulnerabilities...
2020-07-14T16:23:17.156+0700    INFO    Detecting bundler vulnerabilities...

Gemfile.lock
============
Total: 0 (HIGH: 0, CRITICAL: 0)


yarn.lock
=========
Total: 1 (HIGH: 1, CRITICAL: 0)

+---------+------------------+----------+-------------------+---------------+--------------------------------+
| LIBRARY | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |             TITLE              |
+---------+------------------+----------+-------------------+---------------+--------------------------------+
| lodash  | NSWG-ECO-516     | HIGH     | 4.17.15           |               | Allocation of Resources        |
|         |                  |          |                   |               | Without Limits or Throttling   |
+---------+------------------+----------+-------------------+---------------+--------------------------------+
```

2. With no severity severity filter
```
go run cmd/trivy/main.go repo https://github.com/tootsuite/mastodon         
Enumerating objects: 38799, done.
Counting objects: 100% (38799/38799), done.
Compressing objects: 100% (20542/20542), done.
Total 38799 (delta 22694), reused 32144 (delta 17239), pack-reused 0
2020-07-14T16:51:47.904+0700    WARN    OS is not detected and vulnerabilities in OS packages are not detected.
2020-07-14T16:51:47.905+0700    INFO    Detecting yarn vulnerabilities...
2020-07-14T16:51:47.926+0700    INFO    Detecting bundler vulnerabilities...

Gemfile.lock
============
Total: 2 (UNKNOWN: 0, LOW: 0, MEDIUM: 2, HIGH: 0, CRITICAL: 0)

+-----------+------------------+----------+-------------------+--------------------------------+--------------------------------+
|  LIBRARY  | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |         FIXED VERSION          |             TITLE              |
+-----------+------------------+----------+-------------------+--------------------------------+--------------------------------+
| omniauth  | CVE-2015-9284    | MEDIUM   | 1.9.1             |                                | rubygem-omniauth: request      |
|           |                  |          |                   |                                | phase of the OmniAuth Ruby gem |
|           |                  |          |                   |                                | is vulnerable to Cross-Site... |
+-----------+------------------+          +-------------------+--------------------------------+--------------------------------+
| sprockets | CVE-2014-7819    |          | 3.7.2             | 2.7.1, 2.12.3, 2.11.3, 2.10.2, | rubygem-sprockets: arbitrary   |
|           |                  |          |                   | 2.9.4, 2.8.3, 2.5.1, 2.4.6,    | file existence disclosure      |
|           |                  |          |                   | 2.2.3, 2.1.4, 2.0.5            |                                |
+-----------+------------------+----------+-------------------+--------------------------------+--------------------------------+

yarn.lock
=========
Total: 3 (UNKNOWN: 0, LOW: 0, MEDIUM: 2, HIGH: 1, CRITICAL: 0)

+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
| LIBRARY  |  VULNERABILITY ID   | SEVERITY | INSTALLED VERSION |    FIXED VERSION    |             TITLE              |
+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
| acorn    | GHSA-6chw-6frg-f759 | MEDIUM   | 5.7.3             | 5.7.4, 7.1.1, 6.4.1 | Regular Expression Denial of   |
|          |                     |          |                   |                     | Service in Acorn               |
+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
| lodash   | NSWG-ECO-516        | HIGH     | 4.17.15           |                     | Allocation of Resources        |
|          |                     |          |                   |                     | Without Limits or Throttling   |
+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
| minimist | CVE-2020-7598       | MEDIUM   | 1.1.3             | 1.2.3, 0.2.1        | nodejs-minimist: prototype     |
|          |                     |          |                   |                     | pollution allows adding        |
|          |                     |          |                   |                     | or modifying properties of     |
|          |                     |          |                   |                     | Object.prototype using a...    |
+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
```
3. With severity filter `MEDIUM,HIGH`
```
go run cmd/trivy/main.go repo --severity=MEDIUM,HIGH https://github.com/tootsuite/mastodon        
Enumerating objects: 38799, done.
Counting objects: 100% (38799/38799), done.
Compressing objects: 100% (20542/20542), done.
Total 38799 (delta 22695), reused 32144 (delta 17239), pack-reused 0
2020-07-14T16:54:14.913+0700    WARN    OS is not detected and vulnerabilities in OS packages are not detected.
2020-07-14T16:54:14.913+0700    INFO    Detecting bundler vulnerabilities...
2020-07-14T16:54:14.936+0700    INFO    Detecting yarn vulnerabilities...

Gemfile.lock
============
Total: 2 (MEDIUM: 2, HIGH: 0)

+-----------+------------------+----------+-------------------+--------------------------------+--------------------------------+
|  LIBRARY  | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |         FIXED VERSION          |             TITLE              |
+-----------+------------------+----------+-------------------+--------------------------------+--------------------------------+
| omniauth  | CVE-2015-9284    | MEDIUM   | 1.9.1             |                                | rubygem-omniauth: request      |
|           |                  |          |                   |                                | phase of the OmniAuth Ruby gem |
|           |                  |          |                   |                                | is vulnerable to Cross-Site... |
+-----------+------------------+          +-------------------+--------------------------------+--------------------------------+
| sprockets | CVE-2014-7819    |          | 3.7.2             | 2.7.1, 2.12.3, 2.11.3, 2.10.2, | rubygem-sprockets: arbitrary   |
|           |                  |          |                   | 2.9.4, 2.8.3, 2.5.1, 2.4.6,    | file existence disclosure      |
|           |                  |          |                   | 2.2.3, 2.1.4, 2.0.5            |                                |
+-----------+------------------+----------+-------------------+--------------------------------+--------------------------------+

yarn.lock
=========
Total: 3 (MEDIUM: 2, HIGH: 1)

+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
| LIBRARY  |  VULNERABILITY ID   | SEVERITY | INSTALLED VERSION |    FIXED VERSION    |             TITLE              |
+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
| acorn    | GHSA-6chw-6frg-f759 | MEDIUM   | 5.7.3             | 5.7.4, 7.1.1, 6.4.1 | Regular Expression Denial of   |
|          |                     |          |                   |                     | Service in Acorn               |
+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
| lodash   | NSWG-ECO-516        | HIGH     | 4.17.15           |                     | Allocation of Resources        |
|          |                     |          |                   |                     | Without Limits or Throttling   |
+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
| minimist | CVE-2020-7598       | MEDIUM   | 1.1.3             | 1.2.3, 0.2.1        | nodejs-minimist: prototype     |
|          |                     |          |                   |                     | pollution allows adding        |
|          |                     |          |                   |                     | or modifying properties of     |
|          |                     |          |                   |                     | Object.prototype using a...    |
+----------+---------------------+----------+-------------------+---------------------+--------------------------------+
```

### CODE REVIEW

File: pkg/report/writer.go [hunk 1]
```diff
@@ -116,8 +117,10 @@ func (tw TableWriter) write(result Result) {
 	}
 
 	var results []string
-	for _, severity := range dbTypes.SeverityNames {
-		r := fmt.Sprintf("%s: %d", severity, severityCount[severity])
+
+	for _, severity := range tw.Severities {
```
Comment (knqyf263) [inline pkg/report/writer.go:121]:
```
./trivy image --severity MEDIUM,MEDIUM alpine:3.10.2                                                                              [~/src/github.com/aquasecurity/trivy]
2020-07-20T09:17:02.180+0300    INFO    Detecting Alpine vulnerabilities...

alpine:3.10.2 (alpine 3.10.2)
=============================
Total: 4 (MEDIUM: 4, MEDIUM: 4)
```

Could you unique `tw.Severities`? I recommend defining a new function like `Unique` or `UniqueStrings` in utils.
https://github.com/aquasecurity/trivy/blob/master/pkg/utils/utils.go

```suggestion
	for _, severity := range utils.Unique(tw.Severities) {
```

Comment (h4ckm03d) [reply pkg/report/writer.go]: Oke will do

Comment (h4ckm03d) [reply pkg/report/writer.go]:
```
./trivy image --severity MEDIUM,MEDIUM,MEDIUM,HIGH,HIGH alpine:3.10.2
2020-07-20T19:57:37.220+0700    INFO    Detecting Alpine vulnerabilities...

alpine:3.10.2 (alpine 3.10.2)
=============================
Total: 4 (MEDIUM: 4, HIGH: 0)
```

File: pkg/report/writer.go [hunk 2]
```diff
@@ -116,8 +117,10 @@ func (tw TableWriter) write(result Result) {
 	}
 
 	var results []string
-	for _, severity := range dbTypes.SeverityNames {
-		r := fmt.Sprintf("%s: %d", severity, severityCount[severity])
+
+	for _, severity := range utils.UniqueSeverities(tw.Severities) {
+		severityName := severity.String()
+		r := fmt.Sprintf("%s: %d", severityName, severityCount[severityName])
```
Comment (knqyf263) [inline pkg/report/writer.go:123]: I've come up with a better idea.

```suggestion
	var severities []string
	for _, sev := range tw.Severities {
		severities = append(severities, sev.String())
	}

	for _, severity := range dbTypes.SeverityNames {
		if !utils.StringInSlice(severity, severities) {
			continue
		}
		r := fmt.Sprintf("%s: %d", severity, severityCount[severity])
		results = append(results, r)
	}
```

In this way, we don't need to remove duplicates and sort severities.

File: pkg/utils/utils_test.go [hunk 3]
```diff
@@ -193,3 +194,36 @@ func TestCopyFile(t *testing.T) {
...
+		"many duplicates": {
+			args: args{[]types.Severity{types.SeverityLow, types.SeverityCritical, types.SeverityLow, types.SeverityLow, types.SeverityHigh, types.SeverityHigh}},
+			want: []types.Severity{types.SeverityLow, types.SeverityCritical, types.SeverityHigh},
+		},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			got := UniqueSeverities(tt.args.elements)
+			if diff := cmp.Diff(tt.want, got); diff != "" {
```
[the remark this answers is not in the document]
Comment (h4ckm03d) [reply pkg/utils/utils_test.go:224]: No problem if using `assert.Equal`

File: pkg/report/writer.go (hunk 2 above)
Comment (h4ckm03d) [reply pkg/report/writer.go]: Nice, no need to add new utils

File: go.mod [hunk 4]
```diff
@@ -12,6 +12,7 @@ require (
 	github.com/docker/docker v1.4.2-0.20190924003213-a8608b5b67c7
 	github.com/docker/go-connections v0.4.0
 	github.com/golang/protobuf v1.3.3
+	github.com/google/go-cmp v0.4.0
```
Comment (knqyf263) [inline go.mod:15]: Did you apply `go mod tidy`?

Comment (knqyf263) [reply go.mod]: It's so small that I could do it instead.

**Comment by knqyf263** [comment]
Thanks a lot!

**Pull request merged**